04cpp3/sec02Templates.md

```diff
@@ -0,0 +1,351 @@
+{
+    int numerator = x.getNumerator() * y.getNumerator();
+    int denominator = x.getDenominator() * y.getDenominator();
+    return Fraction(numerator, denominator);
+}
+```
+- This version also overloads the `*` infix operator in the same way. 
+- There is now no member version so we can't call `f1.operator*(f2)`, but we can call the non-member function `operator*(f1, f2)`. 
+- Note that we had to create `get` functions for the private member variables because `operator*` is not a member function in this case and so does not have access to private members. 
+
+## Using Class Members with Templated Types
+
+Writing a templated function which takes type `T` implicitly defines an interface that must be met by `T`. 
+- In the simple adding example we use the `+` operator on objects of type `T`: therefore `T` must implement this operator in order for the function to be successfully generated. 
+- In our vector sampling example we only require that `vector<T>` is a valid type; this places very few restrictions on `T` (although there are some). 
+
+The only restrictions on our templates is that the code is valid once the type substitution has been made. If we have a template parameter `T` then we can access member variables and functions of the class `T`, and this function will be able to be generated for any class which implements those properties. For example, let's take this function:
+```cpp
+template<typename T>
+T& getTheBiggerOne(T &a, T &b)
+{
+    if(a.getArea() >= b.getArea())
+    {
+        return a;
+    }
+    else
+    {
+        return b;
+    }
+}
+```
+- This function will return whichever of its two inputs has the larger area.
+- This is a valid template for any type which implements the member function `getArea()`.
+- Note we don't specify the return type of `getArea()`; for this code to be valid it is only necessary that `>=` is defined for the return type of `getArea()`.
+
+We can call this function using our `Shape` classes (`Shape`, `Circle`, `Square`) from last week, since they implement `getArea()`. But we could also implement a new class, like this one:
+```cpp
+class Country
+{
+    Country(string n, double a, double p) : name(n), area(a), population(p) {}
+
+    double getName() { return name; }
+    double getArea() { return area; }
+    double getPopulation() { return population; }
+
+    private:
+    string name;
+    double area;
+    int population;
+}
+```
+- This class defines countries by their name, area, and population. 
+- This class also fulfills all the conditions of `getTheBiggerOne`:
+    - `getArea()` is implemented.
+    - `>=` is defined for `double`. 
+
+We can use `getTheBiggerOne` with our `Country` class just as well as our `Shape` class, even though they are not (and certainly should not be) related by any inheritance! Templates allow us to define generic code that is broader than inheritance based polymorphism, but comes at the cost of being _static_. 
```

Comment:
```suggestion
We can use `getTheBiggerOne` with our `Country` class just as well as our `Shape` class, even though they are not (and certainly should not be) related by any inheritance! Templates allow us to define generic code that is broader than inheritance based polymorphism, but comes at the cost (or benefit!) of being _static_. 
```

IMO static template checks _can_ be a huge advantage. Plus, if we hint that it could be both, some students might be prompted to wonder what benefits/costs are attached to static compilation/polymorphism.